torch/nn/modules/transformer.py

```diff
@@ -177,15 +184,21 @@ def forward(self, src, mask=None, src_key_padding_mask=None):
         Shape:
             see the docs in Transformer class.
         """
-        output = src
+        if self.batch_first:
+            output = src.permute(1, 0, 2)
+        else:
+            output = src
 
         for mod in self.layers:
             output = mod(output, src_mask=mask, src_key_padding_mask=src_key_padding_mask)
 
         if self.norm is not None:
             output = self.norm(output)
 
-        return output
+        if self.batch_first:
```

Comment:
The reason that we have to permute within `TransformerEncoder` and `TransformerDecoder` is that the encoder/decoder layers don't support `batch_first`. Should we consider to add `batch_first` support to them?

Reply:
This is a good question. It would certainly be less overhead if the encoder/decoder layers themselves have an implementation that natively supports `batch_first` or not. I guess it would be worth figuring out if this permute operation is taking up a significant fraction of the time per forward/backward computation (e.g. >1%)